handbook/python-dataplugins/python-dataplugins.md

```diff
@@ -0,0 +1,310 @@
+that consist of one root, 0...m groups, and 0...n channels:
+
+<figure>
+   <img src="../../img/pydp-tdm_structure.png" width="500" />
+   <figcaption>TDM structure with root, groups and channels</figcaption>
+</figure>
+
+<!-- markdownlint-disable -->
+<details>
+<summary>Example dictionary</summary>
+<!-- markdownlint-enable -->
+
+```python
+self.tdm_tree = {
+   "author": "NI",
+   "description": "Example file",
+   "groups": [{
+         "name": "Example",
+         "description": "First group",
+         "time": datetime.datetime(2020, 2, 11, 15, 31, 59, 342380),
+         "channels": [{
+            "name": "Channel_0",
+            "description": "",
+            "values": [1.2, 1.3, 1.4],
+            "info": "Time in seconds",
+            "type": "DataTypeChnFloat64"
+         }, {
+            "name": "Channel_1",
+            "description": "",
+            "values": [10, 11, 12],
+            "unit_string": "km/h",
+            "type": "DataTypeChnFloat64"
+         }]
+   }]
+}
+
+file_path = os.path.realpath(parameter["file"])
+return {Path(file_path).stem: self.tdm_tree}
+```
+
+</details>
+
+<!-- markdownlint-disable -->
+<details>
+<summary>Dictionary Schema</summary>
+<!-- markdownlint-enable -->
+
+```python
+import datetime
+from schema import And, Schema
+
+Schema({
+      Optional('author'): str,
+      Optional('description'): str,
+      'groups': [{
+         'name': str,
+         Optional('description'): str,
+         Optional('time'): datetime.datetime,
+         'channels': [{
+            'name': str,
+            Optional('description'): str,
+            'values': list,
+            Optional('unit_string'): str,
+            'type': And(str, lambda s: s in (
+               'DataTypeChnFloat32',
+               'DataTypeChnFloat64',
+               'DataTypeChnString',
+               'DataTypeChnDate',
+               'DataTypeChnUInt8',
+               'DataTypeChnInt16',
+               'DataTypeChnInt32',
+               'DataTypeChnInt64'))
+            }]
+      }]}, ignore_extra_keys=True)
+```
+
+All additional "extra keys" will show up as custom properties in DIAdem, Labview
+or SystemLink DataFinder.
+
+</details>
+
+See full example: [csv-read-with-direct-loading](https://github.com/ni/systemlink-operations-handbook/tree/master/examples/python-dataplugin-examples/csv-read-with-direct-loading)
+
+## Callback Loading
+
+When handling a big data set, you might not want to load all data at the same time.
+To ensure the DataPlugin only returns values that applications request:
+
+   1. Assign an empty array to the channel values property:
+
+      ```python
+      ...
+      'groups': [{
+         ...
+         'channels': [{
+            ...
+            'values': [],
+         }]
+      }]
+      ```
+
+   1. Outsource the functionality to load the bulk data values in a separate function.

```

Comment:
1. -> 2. Or did you say this actually appears in the ui, I can't remember...

Reply:
yes it will appear correctly